aten/src/ATen/core/op_registration/op_registration_test.cpp

```diff
@@ -219,4 +220,265 @@ TEST(OperatorRegistrationTest, givenOpWithMultipleKernels_whenKernelsHaveSameDis
+  //   std::vector<std::string>(), [] (const IValue& v) {EXPECT_EQ(0, v.toStringListRef().size());},
+  //   "(str[] a) -> str[]");
+
+
+  // list types (with non-empty list)
+  ArgTypeTestKernel<c10::ArrayRef<double>, std::vector<double>>::test(
+    c10::ArrayRef<double>({1.5, 2.5}), [] (c10::ArrayRef<double> v) {EXPECT_EQ(c10::ArrayRef<double>({1.5, 2.5}), v);},
+    std::vector<double>({3.5, 4.5}), [] (const IValue& v) {EXPECT_EQ(std::vector<double>({3.5, 4.5}), v.toDoubleListRef());});
+  ArgTypeTestKernel<c10::ArrayRef<double>, std::vector<double>>::test(
+    c10::ArrayRef<double>({1.5, 2.5}), [] (c10::ArrayRef<double> v) {EXPECT_EQ(c10::ArrayRef<double>({1.5, 2.5}), v);},
+    std::vector<double>({3.5, 4.5}), [] (const IValue& v) {EXPECT_EQ(std::vector<double>({3.5, 4.5}), v.toDoubleListRef());},
+    "(float[] a) -> float[]");
+  ArgTypeTestKernel<c10::ArrayRef<int64_t>, std::vector<int64_t>>::test(
+    c10::ArrayRef<int64_t>({1, 2}), [] (c10::ArrayRef<int64_t> v) {EXPECT_EQ(c10::ArrayRef<int64_t>({1, 2}), v);},
+    std::vector<int64_t>({3, 4}), [] (const IValue& v) {EXPECT_EQ(std::vector<int64_t>({3, 4}), v.toIntListRef());});
+  ArgTypeTestKernel<c10::ArrayRef<int64_t>, std::vector<int64_t>>::test(
+    c10::ArrayRef<int64_t>({1, 2}), [] (c10::ArrayRef<int64_t> v) {EXPECT_EQ(c10::ArrayRef<int64_t>({1, 2}), v);},
+    std::vector<int64_t>({3, 4}), [] (const IValue& v) {EXPECT_EQ(std::vector<int64_t>({3, 4}), v.toIntListRef());},
+    "(int[] a) -> int[]");
+  // TODO When fixing bool[] and str[] (see above), also add them here
+  ArgTypeTestKernel<c10::ArrayRef<Tensor>, std::vector<Tensor>>::test(
+    c10::ArrayRef<Tensor>({dummyTensor(TensorType1()), dummyTensor(TensorType2())}), [] (c10::ArrayRef<Tensor> v) {
+      EXPECT_EQ(2, v.size());
+      EXPECT_EQ(TensorType1(), v[0].type_id());
+      EXPECT_EQ(TensorType2(), v[1].type_id());
+    },
+    std::vector<Tensor>({dummyTensor(TensorType2()), dummyTensor(TensorType1())}), [] (const IValue& v) {
+      EXPECT_EQ(2, v.toTensorListRef().size());
+      EXPECT_EQ(TensorType2(), v.toTensorListRef()[0].type_id());
+      EXPECT_EQ(TensorType1(), v.toTensorListRef()[1].type_id());
+    });
+  ArgTypeTestKernel<c10::ArrayRef<Tensor>, std::vector<Tensor>>::test(
+    c10::ArrayRef<Tensor>({dummyTensor(TensorType1()), dummyTensor(TensorType2())}), [] (c10::ArrayRef<Tensor> v) {
+      EXPECT_EQ(2, v.size());
+      EXPECT_EQ(TensorType1(), v[0].type_id());
+      EXPECT_EQ(TensorType2(), v[1].type_id());
+    },
+    std::vector<Tensor>({dummyTensor(TensorType2()), dummyTensor(TensorType1())}), [] (const IValue& v) {
+      EXPECT_EQ(2, v.toTensorListRef().size());
+      EXPECT_EQ(TensorType2(), v.toTensorListRef()[0].type_id());
+      EXPECT_EQ(TensorType1(), v.toTensorListRef()[1].type_id());
+    },
+    "(Tensor[] a) -> Tensor[]");
+
+
+  // TODO Do we want to support list of optional / optional of list ?
+}
```

Comment:
what about dicts? @driazati - I'm not sure whether you added them to custom ops or not

Reply:
I did, `std::unordered_map` will convert to `c10::ivalue::GenericDict` (and so will `std::vector` to `GenericList`) if it's a custom op argument (#17587)

Reply:
I'm not sure if we want to support `std::vector` arguments. Taking a `std::vector` argument doesn't have any advantage over taking an `ArrayRef` argument, but it prevents us internally from doing optimizations like representing lists as `SmallVector` if we decide to do so. Can we revert or deprecate the `std::vector` support?

For dict, with the same argument as above, we could advocate for passing in a `DictRef` class we'd have to write. Not sure if that's worth the effort.

Reply:
It might be fine to break std::vector support - I don't think it's used that widely. But we'd need to add it in release notes.

As for dicts - I think we can just say that users have to use particular typedef? then we could change this typedef later. But unordered_map will likely leak anyway...

Theoretically, we could have "schema inference from c++ function" to remember what type of the argument it is and dispatch to the correct lowering code. So we potentially do have path for future optimizations. I'd start now with just going with unordered_map and not overthink it too much.

My original comment was that we do want to add tests for dicts here as well

Reply:
Virtualized DictRef sounds unwise. Non-ABI but source recompile compatibility with a typedef/using seems better. We can just guarantee that any replacement implements `unordered_map` equivalent interface (which is how all the replacements work anyway.)

Reply:
I guess we could wrap an unordered_map into a wrapper class with the same interface. This way, we could exchange the implementation without users noticing. With `using`, it would still not be source compatible if we exchange it because users would do

    std::unordered_map<...> a = op();

which would break if `op()` returns a different class with the same interface.